README.md
# Jalaliday
[![npm (scoped with tag)](https://img.shields.io/npm/v/jalaliday/latest.svg?style=flat-square)](https://npmjs.com/package/jalaliday)
[![npm](https://img.shields.io/npm/dt/jalaliday.svg?style=flat-square)](https://npmjs.com/package/jalaliday)
[![CircleCI](https://img.shields.io/circleci/project/github/alibaba-aero/jalaliday.svg?style=flat-square)](https://circleci.com/gh/)
[![Codecov](https://img.shields.io/codecov/c/github/alibaba-aero/jalaliday.svg?style=flat-square)](https://codecov.io/gh/)
[![Dependencies](https://david-dm.org/jalaliday/status.svg?style=flat-square)](https://david-dm.org/)
[![js-standard-style](https://img.shields.io/badge/code_style-standard-brightgreen.svg?style=flat-square)](http://standardjs.com)

Persian (Jalali, Khorshidi) Plugin for [Day.js](https://github.com/iamkun/dayjs), Jalaliday add multi-calendar functionality to Day.js core regardless for of locale, so we can have Gregorian calendar is Persian locale of Jalali calendar in English locale  
Unlike moment and becuase of immutablity of dayjs, there is no need for formats like `jYYYY` or `jMM`, in Jalaliday all formats are same and standard

> 

[📖 **Release Notes**](./CHANGELOG.md)

## Installation
NPM
```
npm install --save jalaliday
```
YARN
```
yarn add jalaliday
```

## Usage
```javascript
import dayjs from 'dayjs'
import jalaliday from 'jalaliday'

dayjs.extend(jalaliday)
```

### Changing calendar
If you want to all new instanses of dayjs use `jalali` calendar, you can set default calendar
```javascript
dayjs.calendar('jalali') // Jalali Calendar
// OR
dayjs.calendar('gregory') // Gregorian Calendar
```
also you can create a jalali date without changing default calendar
```javascript
const date = dayjs()
const jalaliDate = date.calendar('jalali')
```

### Parse Date
- Parse Gregory date 
```js
const date = dayjs('2018-04-04T16:00:00.000Z');
```
- Parse Jalali date
```js
const date = dayjs('1398-10-17', { jalali: true });
```

### Multiple Locale
with combination of `calendar` and `locale` we have multi language for real
```javascript
dayjs().calendar('jalali').locale('en').format('DD MMMM YYYY') // '13 Shahrivar 1397'
dayjs().calendar('gregory').locale('fa').format('DD MMMM YYYY') // '04 سپتامبر 2018'
```

### API
All Api operations of Jalaliday is same as Dayjs itself but calendar based, for more information checkout [Dayjs API](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md)
For a glance:
- [API Reference](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#api-reference)
  - [Parsing](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#parsing)
    - [Constructor `dayjs(existing?: string | number | Date | Dayjs)`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#constructor-dayjsexisting-string--number--date--dayjs)
      - [ISO 8601 string](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#iso-8601-string)
      - [Unix Timestamp (milliseconds since the Unix Epoch - Jan 1 1970, 12AM UTC)](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#unix-timestamp-milliseconds-since-the-unix-epoch---jan-1-1970-12am-utc)
      - [Native Javascript Date object](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#native-javascript-date-object)
    - [Clone `.clone() | dayjs(original: Dayjs)`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#clone-clone-dayjsoriginal-dayjs)
    - [Validation `.isValid()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#validation-isvalid)
  - [Get and Set](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#get-and-set)
    - [Year `.year()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#year-year)
    - [Month `.month()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#month-month)
    - [Day of the Month `.date()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#day-of-the-month-date)
    - [Day of the Week `.day()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#day-of-the-week-day)
    - [Hour `.hour()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#hour-hour)
    - [Minute `.minute()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#minute-minute)
    - [Second `.second()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#second-second)
    - [Millisecond `.millisecond()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#millisecond-millisecond)
    - [Set `.set(unit: string, value: number)`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#set-setunit-string-value-number)
  - [Manipulating](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#manipulating)
    - [Add `.add(value: number, unit: string)`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#add-addvalue-number-unit-string)
    - [Subtract `.subtract(value: number, unit: string)`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#subtract-subtractvalue-number-unit-string)
    - [Start of Time `.startOf(unit: string)`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#start-of-time-startofunit-string)
    - [End of Time `.endOf(unit: string)`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#end-of-time-endofunit-string)
  - [Displaying](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#displaying)
    - [Format `.format(stringWithTokens: string)`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#format-formatstringwithtokens-string)
      - [List of all available formats](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#list-of-all-available-formats)
    - [Difference `.diff(compared: Dayjs, unit: string (default: 'milliseconds'), float?: boolean)`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#difference-diffcompared-dayjs-unit-string-default-milliseconds-float-boolean)
    - [Unix Timestamp (milliseconds) `.valueOf()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#unix-timestamp-milliseconds-valueof)
    - [Unix Timestamp (seconds) `.unix()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#unix-timestamp-seconds-unix)
    - [Days in the Month `.daysInMonth()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#days-in-the-month-daysinmonth)
    - [As Javascript Date `.toDate()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#as-javascript-date-todate)
    - [As Array `.toArray()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#as-array-toarray)
    - [As JSON `.toJSON()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#as-json-tojson)
    - [As ISO 8601 String `.toISOString()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#as-iso-8601-string-toisostring)
    - [As Object `.toObject()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#as-object-toobject)
    - [As String `.toString()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#as-string-tostring)
  - [Query](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#query)
    - [Is Before `.isBefore(compared: Dayjs)`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#is-before-isbeforecompared-dayjs)
    - [Is Same `.isSame(compared: Dayjs)`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#is-same-issamecompared-dayjs)
    - [Is After `.isAfter(compared: Dayjs)`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#is-after-isaftercompared-dayjs)
    - [Is a Dayjs `.isDayjs()`](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#is-a-dayjs-isdayjscompared-any)
  - [Plugin APIs](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#plugin-apis)
    - [RelativeTime](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#relativetime)
    - [IsLeapYear](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#isleapyear)
    - [WeekOfYear](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#weekofyear)
    - [IsBetween](https://github.com/iamkun/dayjs/blob/master/docs/en/API-reference.md#isbetween)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.3.0](https://github.com/alibaba-aero/jalaliday/compare/v2.2.8...v2.3.0) (2020-02-29)


### Features

* **weekday:** support weekday plugin ([#24](https://github.com/alibaba-aero/jalaliday/issues/24)) ([5b29733](https://github.com/alibaba-aero/jalaliday/commit/5b29733))

### [2.2.8](https://github.com/alibaba-aero/jalaliday/compare/v2.2.7...v2.2.8) (2020-02-01)


### Bug Fixes

* **utc:** add utc property to wrapper ([c9e83ca](https://github.com/alibaba-aero/jalaliday/commit/c9e83ca))
* **utc:** add utc property to wrapper ([#22](https://github.com/alibaba-aero/jalaliday/issues/22)) ([92dba28](https://github.com/alibaba-aero/jalaliday/commit/92dba28))

### [2.2.7](https://github.com/alibaba-aero/jalaliday/compare/v2.2.6...v2.2.7) (2020-01-26)

### [2.2.6](https://github.com/alibaba-aero/jalaliday/compare/v2.2.5...v2.2.6) (2020-01-26)


### Bug Fixes

* **safari:** add one hour to startOf to prevent safari issue ([9c512ce](https://github.com/alibaba-aero/jalaliday/commit/9c512ce))

## [2.2.5](https://github.com/alibaba-aero/jalaliday/compare/v2.2.4...v2.2.5) (2019-10-31)


### Bug Fixes

* **$set:** prevent date from exceeding month limit ([#19](https://github.com/alibaba-aero/jalaliday/issues/19)) ([66c8e76](https://github.com/alibaba-aero/jalaliday/commit/66c8e76))



## [2.2.5](https://github.com/alibaba-aero/jalaliday/compare/v2.2.4...v2.2.5) (2019-10-31)


### Bug Fixes

* **$set:** prevent date from exceeding month limit ([#19](https://github.com/alibaba-aero/jalaliday/issues/19)) ([66c8e76](https://github.com/alibaba-aero/jalaliday/commit/66c8e76))



## [2.2.4](https://github.com/alibaba-aero/jalaliday/compare/v2.2.3...v2.2.4) (2019-09-22)


### Bug Fixes

* **add-month:** prevent exceed from days of month ([9835799](https://github.com/alibaba-aero/jalaliday/commit/9835799))

### [2.2.3](https://github.com/alibaba-aero/jalaliday/compare/v2.2.2...v2.2.3) (2019-08-23)


### Bug Fixes

* support dayjs 1.8.15 ([cc5754c](https://github.com/alibaba-aero/jalaliday/commit/cc5754c))

## [2.2.2](https://github.com/alibaba-aero/jalaliday/compare/v2.2.1...v2.2.2) (2019-03-14)


### Bug Fixes

* **toArray:** override toArray if exists ([00c6548](https://github.com/alibaba-aero/jalaliday/commit/00c6548))



## [2.2.1](https://github.com/alibaba-aero/jalaliday/compare/v2.2.0...v2.2.1) (2019-03-10)


### Bug Fixes

* **plugin:** refactor add and startOf functionality ([8435a56](https://github.com/alibaba-aero/jalaliday/commit/8435a56)), closes [#8](https://github.com/alibaba-aero/jalaliday/issues/8)
* **plugin:** support dayjs^1.8.9 ([31bd189](https://github.com/alibaba-aero/jalaliday/commit/31bd189)), closes [#6](https://github.com/alibaba-aero/jalaliday/issues/6)



<a name="2.2.0"></a>
# [2.2.0](https://github.com/alibaba-aero/jalaliday/compare/v2.1.1...v2.2.0) (2019-02-27)


### Features

* support toArray proto for jalali calendar ([e7b9c9e](https://github.com/alibaba-aero/jalaliday/commit/e7b9c9e))



<a name="2.1.1"></a>
## [2.1.1](https://github.com/alibaba-aero/jalaliday/compare/v2.0.1...v2.1.1) (2019-02-02)


### Bug Fixes

* **installation command:** changes yard to yarn ([13eec41](https://github.com/alibaba-aero/jalaliday/commit/13eec41))
* **linter errors:** fixes linter errors on plugin, reliable and test files ([a209460](https://github.com/alibaba-aero/jalaliday/commit/a209460))
* **relativeTimes:** adds the object to the fa constant ([e280f2a](https://github.com/alibaba-aero/jalaliday/commit/e280f2a))


### Features

* **relative times:** adds relative time strings for fa locale ([e961a1f](https://github.com/alibaba-aero/jalaliday/commit/e961a1f))



<a name="2.0.1"></a>
## [2.0.1](https://github.com/alibaba-aero/jalaliday/compare/v2.0.0...v2.0.1) (2018-12-07)


### Bug Fixes

* wrap dayjs functions to keep calendar on new instances ([f10e9fe](https://github.com/alibaba-aero/jalaliday/commit/f10e9fe))



<a name="1.1.1"></a>
## [1.1.1](https://github.com/alibaba-aero/jalaliday/compare/v1.1.0...v1.1.1) (2018-12-05)


### Bug Fixes

* **date:** rename day function to date ([dfed41e](https://github.com/alibaba-aero/jalaliday/commit/dfed41e))
* **utils.js:** Fix typo in monthDiff function ([df0678e](https://github.com/alibaba-aero/jalaliday/commit/df0678e))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/alibaba-aero/jalaliday/compare/v1.0.0...v1.1.0) (2018-09-02)


### Features

* **startOf:** Support customizable first day of week ([6e3dac4](https://github.com/alibaba-aero/jalaliday/commit/6e3dac4))



<a name="1.0.0"></a>
# 1.0.0 (2018-09-01)


test/plugin.test.js
import MockDate from 'mockdate'
import dayjs from 'dayjs'
import jalali from '../src'

dayjs.extend(jalali)
dayjs.calendar('jalali')

beforeEach(() => {
  MockDate.set(new Date())
})

afterEach(() => {
  MockDate.reset()
})

it('Extend dayjs', () => {
  expect(dayjs.$C).toBeDefined()
  expect(dayjs.$C).toEqual('jalali')
  expect(dayjs().$jy).toBeDefined()
})

it('Setting calendar converts date', () => {
  const date = dayjs('1397/06/13', { jalali: true }).calendar('gregory')
  expect(date.$y).toEqual(2018)
  expect(date.$M).toEqual(8)
  expect(date.$D).toEqual(4)

  const date2 = dayjs('2018/09/04').calendar('jalali')
  expect(date2.$jy).toEqual(1397)
  expect(date2.$jM).toEqual(5)
  expect(date2.$jD).toEqual(13)
})

test('keep instance calendar on manipulation', () => {
  const date = dayjs().calendar('jalali')
  expect(date.add(1, 'month').isJalali()).toEqual(true)
  expect(dayjs(date).startOf('month').$C).toEqual('jalali')
  expect(dayjs(date).add(1, 'month').$C).toEqual('jalali')
  expect(dayjs(date).add(1, 'month').add(1, 'month').isJalali()).toEqual(true)
})


test/plugin.add.test.js
import MockDate from 'mockdate'
import dayjs from 'dayjs'
import jalali from '../src'

dayjs.extend(jalali)
dayjs.calendar('jalali')

beforeEach(() => {
  MockDate.set(new Date())
})

afterEach(() => {
  MockDate.reset()
})

it('add 1 day in the middle of the month', () => {
  const date = dayjs('1397/06/13', { jalali: true })
  const date2 = date.add(1, 'day')
  expect(date2.year()).toEqual(date.year())
  expect(date2.month()).toEqual(date.month())
  expect(date2.date()).toEqual(date.date() + 1)
})

it('add 1 day in the end of the month', () => {
  const date = dayjs('1397/06/31', { jalali: true })
  const date2 = date.add(1, 'day')
  expect(date2.year()).toEqual(date.year())
  expect(date2.month()).toEqual(date.month() + 1)
  expect(date2.date()).toEqual(1)
})

it('add 1 month in the middle of the year', () => {
  const date = dayjs('1397/06/13', { jalali: true })
  const date2 = date.add(1, 'month')
  expect(date2.year()).toEqual(date.year())
  expect(date2.month()).toEqual(date.month() + 1)
  expect(date2.date()).toEqual(date.date())
})

it('add 1 month in the end of the year', () => {
  const date = dayjs('1397/12/13', { jalali: true })
  const date2 = date.add(1, 'month')
  expect(date2.year()).toEqual(date.year() + 1)
  expect(date2.month()).toEqual(0)
  expect(date2.date()).toEqual(date.date())
})

it('add 1 year', () => {
  const date = dayjs('1397/12/13', { jalali: true })
  const date2 = date.add(1, 'year')
  expect(date2.year()).toEqual(date.year() + 1)
  expect(date2.month()).toEqual(date.month())
  expect(date2.date()).toEqual(date.date())
})

test('add 11 months in the middle of the year', () => {
  const date = dayjs('1396/06/13', { jalali: true })
  const date2 = date.add(11, 'month')
  expect(date2.year()).toEqual(date.year() + 1)
  expect(date2.month()).toEqual(date.month() - 1)
  expect(date2.date()).toEqual(date.date())
})

describe('add 100 days', () => {
  let a = null
  let b = null

  beforeEach(() => {
    a = dayjs('1397/06/01', { jalali: true })
    b = dayjs('1397/09/10', { jalali: true })
  })

  it('add date', () => {
    expect(a.add(100, 'day')).toEqual(b)
  })
})

describe('add 1 month in shahrivar 31th', () => {
  let a = null
  let b = null

  beforeEach(() => {
    a = dayjs('1397/06/31', { jalali: true })
    b = dayjs('1397/07/30', { jalali: true })
  })

  it('add date', () => {
    expect(a.add(1, 'month')).toEqual(b)
  })
})


test/plugin.subtract.js
import MockDate from 'mockdate'
import dayjs from 'dayjs'
import jalali from '../src'

dayjs.extend(jalali)
dayjs.calendar('jalali')

beforeEach(() => {
  MockDate.set(new Date())
})

afterEach(() => {
  MockDate.reset()
})

test('subtract 1 ms', () => {
  const date = dayjs('1397/06/13', { jalali: true })
  const date2 = date.subtract(1, 'ms')
  expect(date2.year()).toEqual(date.year())
  expect(date2.month()).toEqual(date.month())
  expect(date2.date()).toEqual(date.date() - 1)
  expect(date2.$ms).toEqual(999)
})

test('subtract 1 day in the middle of the month', () => {
  const date = dayjs('1397/06/13', { jalali: true })
  const date2 = date.subtract(1, 'day')
  expect(date2.year()).toEqual(date.year())
  expect(date2.month()).toEqual(date.month())
  expect(date2.date()).toEqual(date.date() - 1)
})

test('subtract 1 day in the beginning of the month', () => {
  const date = dayjs('1397/06/01', { jalali: true })
  const date2 = date.subtract(1, 'day')
  expect(date2.year()).toEqual(date.year())
  expect(date2.month()).toEqual(date.month() - 1)
  expect(date2.date()).toEqual(31)
})

test('subtract 1 month in the middle of the year', () => {
  const date = dayjs('1397/06/13', { jalali: true })
  const date2 = date.subtract(1, 'month')
  expect(date2.year()).toEqual(date.year())
  expect(date2.month()).toEqual(date.month() - 1)
  expect(date2.date()).toEqual(date.date())
})


test('subtract 18 months', () => {
  const date = dayjs('1397/06/13', { jalali: true })
  const date2 = date.subtract(18, 'month')
  expect(date2.year()).toEqual(1395)
  expect(date2.month()).toEqual(11)
  expect(date2.date()).toEqual(date.date())
})

test('subtract 17 months', () => {
  const date = dayjs('1397/06/13', { jalali: true })
  const date2 = date.subtract(17, 'month')
  expect(date2.year()).toEqual(1396)
  expect(date2.month()).toEqual(0)
  expect(date2.date()).toEqual(date.date())
})

test('subtract 1 month in the beginning of the year', () => {
  const date = dayjs('1397/01/13', { jalali: true })
  const date2 = date.subtract(1, 'month')
  expect(date2.year()).toEqual(date.year() - 1)
  expect(date2.month()).toEqual(11)
  expect(date2.date()).toEqual(date.date())
})

test('subtract 1 month in the beginning of the year', () => {
  const date = dayjs('1397/01/01', { jalali: true })
  const date2 = date.subtract(11, 'month')
  expect(date2.year()).toEqual(date.year() - 1)
  expect(date2.month()).toEqual(1)
  expect(date2.date()).toEqual(date.date())
})

test('subtract 1 year', () => {
  const date = dayjs('1397/01/13', { jalali: true })
  const date2 = date.subtract(1, 'year')
  expect(date2.year()).toEqual(date.year() - 1)
  expect(date2.month()).toEqual(date.month())
  expect(date2.date()).toEqual(date.date())
})


src/utils.js
const padStart = (string, length, pad) => {
  const s = String(string)
  if (!s || s.length >= length) return string
  return `${Array((length + 1) - s.length).join(pad)}${string}`
}

const monthDiff = (a, b) => {
  // function from moment.js in order to keep the same result
  const wholeMonthDiff = ((b.year() - a.year()) * 12) + (b.month() - a.month())
  const anchor = a.clone().add(wholeMonthDiff, 'months')
  const c = b - anchor < 0
  const anchor2 = a.clone().add(wholeMonthDiff + (c ? -1 : 1), 'months')
  return Number(-(wholeMonthDiff + ((b - anchor) / (c ? (anchor - anchor2) : (anchor2 - anchor)))))
}

const absFloor = n => (n < 0 ? Math.ceil(n) || 0 : Math.floor(n))

const prettyUnit = u => (u && String(u).toLowerCase().replace(/s$/, ''))

const isUndefined = s => s === void 0 // eslint-disable-line no-void

export default {
  padStart,
  monthDiff,
  absFloor,
  prettyUnit,
  isUndefined
}



src/plugin.js
import fa from 'dayjs/locale/fa'

import jdate from './calendar'
import * as C from './constant'

export default (o, Dayjs, dayjs) => {
  const proto = Dayjs.prototype
  const U = proto.$utils()
  const $isJalali = (v) => v.$C === 'jalali'
  const $prettyUnit = U.prettyUnit || U.p
  const $isUndefined = U.isUndefined || U.u
  const $padStart = U.padStart || U.s
  const $monthDiff = U.monthDiff || U.m
  const $absFloor = U.absFloor || U.a
  const wrapperOfTruth = (action) => function (...args) {
    const unsure = action.bind(this)(...args)
    unsure.$C = this.$C
    if (unsure.isJalali()) {
      unsure.InitJalali()
    }
    return unsure
  }

  // keep calendar on date manipulation
  proto.startOf = wrapperOfTruth(proto.startOf)
  proto.endOf = wrapperOfTruth(proto.endOf)
  proto.add = wrapperOfTruth(proto.add)
  proto.subtract = wrapperOfTruth(proto.subtract)
  proto.set = wrapperOfTruth(proto.set)
  const oldParse = proto.parse
  const oldInit = proto.init
  const oldStartOf = proto.startOf
  const old$Set = proto.$set
  const oldAdd = proto.add
  const oldFormat = proto.format
  const oldDiff = proto.diff
  const oldYear = proto.year
  const oldMonth = proto.month
  const oldDate = proto.date
  const oldDaysInMonth = proto.daysInMonth
  const oldToArray = proto.toArray

  dayjs.$C = 'gregory'
  // First Day Of Week
  dayjs.$fdow = 6 // 0: sunday, ...

  dayjs.calendar = function (calendar) {
    dayjs.$C = calendar
    return dayjs
  }

  proto.calendar = function (calendar) {
    const that = this.clone()
    that.$C = calendar
    if (that.isJalali()) {
      that.InitJalali()
    }
    return that
  }

  proto.isJalali = function () {
    return $isJalali(this)
  }

  dayjs.en.jmonths = 'Farvardin_Ordibehesht_Khordaad_Tir_Mordaad_Shahrivar_Mehr_Aabaan_Aazar_Dey_Bahman_Esfand'.split('_')
  dayjs.locale('fa', { ...fa, ...C.fa }, true)

  const wrapper = function (date, instance) {
    return dayjs(date, {
      locale: instance.$L,
      utc: instance.$u,
      calendar: instance.$C
    })
  }

  proto.init = function (cfg = {}) {
    oldInit.bind(this)(cfg)

    if (this.isJalali()) {
      this.InitJalali()
    }
  }

  proto.parse = function (cfg) {
    let reg
    this.$C = cfg.calendar || this.$C || dayjs.$C
    // eslint-disable-next-line no-cond-assign
    if (cfg.jalali && (typeof cfg.date === 'string')
      && (/.*[^Z]$/i.test(cfg.date)) // looking for a better way
      && (reg = cfg.date.match(C.REGEX_PARSE))) {
      // 1397-08-08 or 13970808
      const [y, m, d] = jdate.G(
        parseInt(reg[1], 10),
        parseInt(reg[2], 10),
        parseInt(reg[3] || 1, 10)
      )
      cfg.date = `${y}-${m}-${d}${reg[4] || ''}`
    }
    return oldParse.bind(this)(cfg)
  }

  proto.InitJalali = function () {
    const [jy, jm, jd] = jdate.J(this.$y, this.$M + 1, this.$D)
    this.$jy = jy
    this.$jM = jm - 1
    this.$jD = jd
  }

  proto.startOf = function (units, startOf) { // startOf -> endOf
    if (!$isJalali(this)) {
      return oldStartOf.bind(this)(units, startOf)
    }
    const isStartOf = !$isUndefined(startOf) ? startOf : true
    const unit = $prettyUnit(units)
    const instanceFactory = (d, m, y = this.$jy) => {
      const [gy, gm, gd] = jdate.G(y, m + 1, d)
      const ins = wrapper(new Date(gy, gm - 1, gd), this)
      return (isStartOf ? ins : ins.endOf(C.D))
        .$set('hour', 1) // prevent daylight saving issue in safari
    }
    const WModifier = (this.$W + (7 - dayjs.$fdow)) % 7
    switch (unit) {
      case C.Y:
        return isStartOf ? instanceFactory(1, 0)
          : instanceFactory(0, 0, this.$jy + 1)
      case C.M:
        return isStartOf ? instanceFactory(1, this.$jM)
          : instanceFactory(
            0,
            (this.$jM + 1) % 12,
            this.$jy + parseInt((this.$jM + 1) / 12, 10)
          )
      case C.W:
        return isStartOf ? instanceFactory(this.$jD - WModifier, this.$jM)
          : instanceFactory(this.$jD + (6 - WModifier), this.$jM)
      default:
        return oldStartOf.bind(this)(units, startOf)
    }
  }

  proto.$set = function (units, int) {
    if (!$isJalali(this)) {
      return old$Set.bind(this)(units, int)
    }
    const unit = $prettyUnit(units)
    const instanceFactory = (d, m, y = this.$jy) => {
      const [gy, gm, gd] = jdate.G(y, m + 1, d)
      this.$d.setFullYear(gy)
      this.$d.setMonth(gm - 1)
      this.$d.setDate(gd)
      return this
    }
    switch (unit) {
      case C.DATE:
      case C.D:
        instanceFactory(int, this.$jM)
        break
      case C.M:
        instanceFactory(this.$jD, int)
        break
      case C.Y:
        instanceFactory(this.$jD, this.$jM, int)
        break
      default:
        return old$Set.bind(this)(units, int)
    }
    this.init()
    return this
  }

  proto.add = function (number, units) {
    if (!$isJalali(this)) {
      return oldAdd.bind(this)(number, units)
    }
    number = Number(number) // eslint-disable-line no-param-reassign
    // units === 'ms' hard code here, will update in next release
    const unit = (units && (units.length === 1 || units === 'ms')) ? units : $prettyUnit(units)
    const instanceFactory = (u, n) => {
      const date = this.set(C.DATE, 1).set(u, n + number)
      return date.set(C.DATE, Math.min(this.$jD, date.daysInMonth()))
    }
    if (['M', C.M].indexOf(unit) > -1) {
      const n = this.$jM + number
      const y = n < 0 ? -Math.ceil(-n / 12) : parseInt(n / 12, 10)
      const d = this.$jD
      const x = this.set(C.D, 1).add(y, C.Y).set(C.M, n - (y * 12))
      return x.set(C.D, Math.min(x.daysInMonth(), d))
    }
    if (['y', C.Y].indexOf(unit) > -1) {
      return instanceFactory(C.Y, this.$jy)
    }
    if (['d', C.D].indexOf(unit) > -1) {
      const date = new Date(this.$d)
      date.setDate(date.getDate() + number)
      return wrapper(date, this)
    }

    return oldAdd.bind(this)(number, units)
  }

  proto.format = function (formatStr, localeObject) {
    if (!$isJalali(this)) {
      return oldFormat.bind(this)(formatStr, localeObject)
    }
    const str = formatStr || C.FORMAT_DEFAULT
    const locale = localeObject || this.$locale()
    const { jmonths } = locale
    return str.replace(C.REGEX_FORMAT, (match) => {
      if (match.indexOf('[') > -1) return match.replace(/\[|\]/g, '')
      switch (match) {
        case 'YY':
          return String(this.$jy).slice(-2)
        case 'YYYY':
          return String(this.$jy)
        case 'M':
          return String(this.$jM + 1)
        case 'MM':
          return $padStart(this.$jM + 1, 2, '0')
        case 'MMM':
          return jmonths[this.$jM].slice(0, 3)
        case 'MMMM':
          return jmonths[this.$jM]
        case 'D':
          return String(this.$jD)
        case 'DD':
          return $padStart(this.$jD, 2, '0')
        default:
          return oldFormat.bind(this)(match, localeObject)
      }
    })
  }

  proto.diff = function (input, units, float) {
    if (!$isJalali(this)) {
      return oldDiff.bind(this)(input, units, float)
    }
    const unit = $prettyUnit(units)
    const that = dayjs(input)
    let result = $monthDiff(this, that)
    switch (unit) {
      case C.Y:
        result /= 12
        break
      case C.M:
        break
      default: // milliseconds
        return oldDiff.bind(this)(input, units, float)
    }
    return float ? result : $absFloor(result)
  }

  proto.$g = function (input, get, set) {
    if ($isUndefined(input)) return this[get]
    return this.set(set, input)
  }

  proto.year = function (input) {
    if (!$isJalali(this)) {
      return oldYear.bind(this)(input)
    }
    return this.$g(input, '$jy', C.Y)
  }

  proto.month = function (input) {
    if (!$isJalali(this)) {
      return oldMonth.bind(this)(input)
    }
    return this.$g(input, '$jM', C.M)
  }

  proto.date = function (input) {
    if (!$isJalali(this)) {
      return oldDate.bind(this)(input)
    }
    return this.$g(input, '$jD', C.D)
  }

  proto.daysInMonth = function () {
    if (!$isJalali(this)) {
      return oldDaysInMonth.bind(this)()
    }
    return this.endOf(C.M).$jD
  }

  /**
   * toArray function moved to official plugin
   * Check function existence before override
   */
  if (oldToArray) {
    proto.toArray = function () {
      if (!$isJalali(this)) {
        return oldToArray.bind(this)()
      }
      return [this.$jy, this.$jM, this.$jD, this.$H, this.$m, this.$s, this.$ms]
    }
  }

  proto.clone = function () {
    return wrapper(this.toDate(), this)
  }
}
